pkg/core/native/notary.go

```diff
@@ -0,0 +1,347 @@
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if till < deposit.Till {
+		return stackitem.NewBool(false)
+	}
+	deposit.Till = till
+	err = n.putDepositFor(ic.DAO, deposit, addr)
+	if err != nil {
+		panic(fmt.Errorf("failed to put deposit for %s into the storage: %w", addr.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+
+// withdraw sends all deposited GAS for "from" address to "to" address.
+func (n *Notary) withdraw(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	from := toUint160(args[0])
+	ok, err := runtime.CheckHashedWitness(ic, from)
+	if err != nil {
+		panic(fmt.Errorf("failed to check witness for %s: %w", from.StringBE(), err))
+	}
+	if !ok {
+		return stackitem.NewBool(false)
+	}
+	to := from
+	if !args[1].Equals(stackitem.Null{}) {
+		to = toUint160(args[1])
+	}
+	deposit := n.getDepositFor(ic.DAO, from)
+	if deposit == nil {
+		return stackitem.NewBool(false)
+	}
+	if ic.Chain.BlockHeight() < deposit.Till {
+		return stackitem.NewBool(false)
+	}
+	err = n.GAS.TransferInternal(ic, n.Hash, to, deposit.Amount)
+	if err != nil {
+		return stackitem.NewBool(false)
+	}
+	if err := n.removeDepositFor(ic.DAO, from); err != nil {
+		panic(fmt.Errorf("failed to remove withdrawn deposit for %s from the storage: %w", from.StringBE(), err))
+	}
+	return stackitem.NewBool(true)
+}
+
+// balanceOf returns deposited GAS amount for specified address.
+func (n *Notary) balanceOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		return stackitem.NewBigInteger(big.NewInt(0))
+	}
+	return stackitem.NewBigInteger(deposit.Amount)
+}
+
+// expirationOf Returns deposit lock height for specified address.
+func (n *Notary) expirationOf(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	acc := toUint160(args[0])
+	deposit := n.getDepositFor(ic.DAO, acc)
+	if deposit == nil {
+		panic(fmt.Errorf("deposit for %s not found", acc.StringBE()))
+	}
+	return stackitem.Make(deposit.Till)
+}
+
+// verify checks whether the transaction was signed by one of the notaries.
+func (n *Notary) verify(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	sig, err := args[0].TryBytes()
+	if err != nil {
+		panic(fmt.Errorf("failed to get signature bytes: %w", err))
+	}
+	tx := ic.Tx
+	if len(tx.GetAttributes(transaction.NotaryAssistedT)) == 0 {
+		return stackitem.NewBool(false)
+	}
+	var (
+		signedByNotary, hasEnoughFunds bool
+		needFee                        *big.Int
+	)
+	for i, signer := range tx.Signers {
+		if signer.Account == n.Hash {
+			signedByNotary = true
+			if signer.Scopes != transaction.None {
+				return stackitem.NewBool(false)
+			}
+			if i == 0 {
+				if len(tx.Signers) != 2 {
+					return stackitem.NewBool(false)
+				}
+				needFee = big.NewInt(tx.NetworkFee + tx.SystemFee)
+			}
+		} else if needFee != nil {
+			balance := n.getDepositFor(ic.DAO, signer.Account)
```

Comment:
`needFee != nil` only if `Notary` is the sender of the transaction. Maybe we can simplify all this loop to checking if `tx.Sender() == n.Hash` and second signer has sufficient deposit?